src/Tablebot/Plugins/Quote.hs

```diff
@@ -47,34 +71,143 @@ quoteComm ::
-quoteComm (WErr (Right (Left (_, qId)))) = showQ (fromIntegral qId)
-quoteComm (WErr (Right (Right (_, qId)))) = deleteQ (fromIntegral qId)
-
--- | @addQuote@, which looks for a message of the form
--- @!quote add "quoted text" - author@, and then stores said quote in the
--- database, returning the ID used.
-addQ :: String -> String -> Message -> DatabaseDiscord ()
-addQ qu author m = do
-  added <- insert $ Quote qu author
-  let res = pack $ show $ fromSqlKey added
-  sendMessage m ("Quote added as #" `append` res)
+quoteComm (WErr (Left (Left (Left (_, Qu qu, _, ROI author))))) = addQ qu author
+quoteComm (WErr (Left (Left (Right (_, qId, Qu qu, _, ROI author))))) = editQ (fromIntegral qId) qu author
+quoteComm (WErr (Left (Right (Left (_))))) = thisQ
+quoteComm (WErr (Left (Right (Right (_, ROI author))))) = authorQ author
+quoteComm (WErr (Right (Left (Left (_, qId))))) = showQ (fromIntegral qId)
+quoteComm (WErr (Right (Left (Right (_, qId))))) = deleteQ (fromIntegral qId)
+quoteComm (WErr (Right (Right (_)))) = randomQ
 
 -- | @showQuote@, which looks for a message of the form @!quote show n@, looks
 -- that quote up in the database and responds with that quote.
 showQ :: Int64 -> Message -> DatabaseDiscord ()
 showQ qId m = do
   qu <- get $ toSqlKey qId
   case qu of
-    Just (Quote txt author) ->
-      sendMessage m $ pack $ txt ++ " - " ++ author
+    Just q -> renderQuoteMessage q qId m
     Nothing -> sendMessage m "Couldn't get that quote!"
 
+-- | @randomQuote@, which looks for a message of the form @!quote random@,
+-- selects a random quote from the database and responds with that quote.
+randomQ :: Message -> DatabaseDiscord ()
+randomQ m = do
+  num <- count allQuotes
+  if num == 0
+    then sendMessage m "Couldn't find any quotes!"
+    else do
+      rindex <- liftIO $ randomRIO (0, (num - 1))
+      key <- selectKeysList allQuotes [OffsetBy rindex, LimitTo 1]
+      qu <- get $ head key
+      case qu of
+        Just q -> renderQuoteMessage q (fromSqlKey $ head key) m
+        Nothing -> sendMessage m "Couldn't find any quotes!"
+      return ()
+  where
+    allQuotes :: [Filter Quote]
+    allQuotes = []
+
+-- | @authorQuote@, which looks for a message of the form @!quote user u@,
+-- selects a random quote from the database attributed to u and responds with that quote.
+-- This is currently mis-documented in the help pages as its not quite working (the author matching is way to strict)
+authorQ :: String -> Message -> DatabaseDiscord ()
+authorQ t m = do
+  num <- count userQuotes
+  if num == 0
+    then sendMessage m "Couldn't find any quotes matching that user!"
+    else do
+      rindex <- liftIO $ randomRIO (0, (num - 1))
+      key <- selectKeysList userQuotes [OffsetBy rindex, LimitTo 1]
+      qu <- get $ head key
+      case qu of
+        Just q -> renderQuoteMessage q (fromSqlKey $ head key) m
+        Nothing -> sendMessage m "Couldn't find any quotes matching that user!"
+      return ()
```

Comment:
Might be worth generalising this to a function `randomFromSet` that takes in the filter condition and does this behaviour? (thus merging general random selection and author-specific random selection, and maybe opening up the possibility of doing time period based selection)